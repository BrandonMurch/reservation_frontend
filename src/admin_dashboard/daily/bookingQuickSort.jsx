/* eslint-disable no-plusplus */
const swap = function swapElements(array, left, right) {
  const temp = array[left];
  array[left] = array[right];
  array[right] = temp;
};

const partition = function swapElementsAroundPivotPoint(array, leftIndex, rightIndex, pivot) {
  let left = leftIndex;
  let right = rightIndex;
  while (left <= right) {
    while (array[left].startTime < pivot) {
      left++;
    }
    while (array[right].startTime > pivot) {
      right--;
    }
    if (left <= right) {
      swap(array, left, right);
      left++;
      right--;
    }
  }
  return left;
};

const sort = function implementQuickSort(array, left, right) {
  if (left >= right) {
    return;
  }

  const pivot = array[Math.floor((left + right) / 2)].startTime;
  const partitionPoint = partition(array, left, right, pivot);
  sort(array, left, partitionPoint - 1);
  sort(array, partitionPoint, right);
};

const quicksort = function quickSortBookingsByStartTime(bookings) {
  sort(bookings, 0, bookings.length - 1);
};

export default quicksort;
