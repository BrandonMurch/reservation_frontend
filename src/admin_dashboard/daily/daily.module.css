@value colors: "../../css/colors.module.css";
@value primary, dark, light, accent from colors;


.container {
   composes: bodyContainer from '../admin_dashboard.module.css';
   background: primary;
}

.tableContainer {
   overflow-x: scroll;
   height: 90%;
}

.table {
   width: 100%;
   padding: 2rem;
}

.tableInput {
   width: 4rem;
   text-align: center;
}

.errorTableInput {
   composes: tableInput;
   background-color: rgba(255, 0, 0, 0.6);
}

.row{
   text-align: center;
}


.hourRow {
   composes: row;
   border: 5px solid black;
}

.errorText {
   color: red;
   font-weight: bold;
   margin: 0.5rem;
}

.editButton {
   padding: 0;
   text-align: center;
}
