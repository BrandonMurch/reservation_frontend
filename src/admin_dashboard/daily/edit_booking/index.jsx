// Dependencies
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { fetchWrapper } from 'shared/useFetch';
import { useTokenContext } from 'contexts/token_context';
import types from './window_types';

// Components
import DeleteConfirmation from './delete_confirmation';
import EditWindow from './edit_window';
import LoadingWindow from './loading_window';
import Exit from './exit_button';

// Stylesheets
import style from './edit_booking.module.css';

const EditBookingOverlay = ({
  booking, exit, refreshBookings, setErrorBanner,
}) => {
  const [windowToDisplay, setWindowToDisplay] = useState(types.EDIT);

  const checkErrorAndExit = function checkResponseForErrorsRefreshBookingsAndExitWindow(
    error, alternativeRender,
  ) {
    if (error) {
      setErrorBanner(alternativeRender);
    } else {
      setErrorBanner(null);
      refreshBookings();
    }

    exit();
  };

  const deleteBooking = async function submitDeleteRequestForBookingToServer() {
    setWindowToDisplay(types.LOADING);
    const { error, alternativeRender } = await fetchWrapper(`/bookings/${booking.id}`, { method: 'DELETE', authorization: `Bearer: ${useTokenContext.getToken}` });
    checkErrorAndExit(error, alternativeRender);
  };

  const submitEdit = async function editBookingOnSubmit(submittedBooking) {
    setWindowToDisplay(types.LOADING);
    const { error, alternativeRender } = await fetchWrapper(
      `/bookings/${booking.id}`, {
        method: 'PUT',
        body: JSON.stringify(submittedBooking),
        authorization: `Bearer: ${useTokenContext.getToken}`,
      },
    );
    checkErrorAndExit(error, alternativeRender);
  };

  let render;
  switch (windowToDisplay) {
    case types.DELETE:
      render = (
        <DeleteConfirmation
          booking={booking}
          cancelDelete={() => setWindowToDisplay(types.EDIT)}
          refreshBookings={refreshBookings}
          deleteBooking={() => deleteBooking(booking)}
        />
      );
      break;

    case types.EDIT:
      render = (
        <EditWindow
          booking={booking}
          onSubmit={(submittedBooking) => submitEdit(submittedBooking)}
          setWindowToDisplay={setWindowToDisplay}
          exit={exit}
        />
      );
      break;

    case types.LOADING:
      render = (
        <LoadingWindow />
      );
      break;

    default:
      render = (
        <h1>
          {'OH NO! This shouldn\'t happen...'}
        </h1>
      );
      break;
  }

  return (
    <div className={style.background}>
      <div className={style.container}>
        <Exit onClick={exit} />
        {render}
      </div>
    </div>
  );
};

EditBookingOverlay.propTypes = {
  exit: PropTypes.func.isRequired,
  refreshBookings: PropTypes.func.isRequired,
  setErrorBanner: PropTypes.func.isRequired,
  booking: PropTypes.shape({
    id: PropTypes.number.isRequired,
  }).isRequired,
};

export default EditBookingOverlay;
