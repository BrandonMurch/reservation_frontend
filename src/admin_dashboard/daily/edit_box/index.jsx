// Dependencies
import React, { useState } from 'react';
import PropTypes from 'prop-types';

// Components
import Form from 'general_components/form';
import DeleteConfirmation from './delete_confirmation';

// Stylesheets
import style from './edit_box.module.css';

const getInputs = function getInputsFromBooking(booking) {
  return [{
    // TODO: get a dropdown of available times?
    name: 'startTime',
    type: 'time',
    label: 'Start Time',
    value: booking.startTime,
    required: true,
  },
  {
    name: 'lastName',
    type: 'time',
    label: 'End Time',
    value: booking.endTime,
    required: true,
  },
  {
    name: 'partySize',
    type: 'text',
    label: 'Party Size',
    value: booking.partySize.toString(),
    required: true,
    pattern: '^[1-9][0-9]*$',
    patternMessage: 'Party size must be a number greater than zero',
  },
  {
    name: 'restaurantComments',
    type: 'text-area',
    label: 'Restaurant Comments',
    value: booking.restaurantComments,
    required: false,
  },
  {
    name: 'userComments',
    type: 'text-area',
    label: 'User Comments',
    value: booking.userComments,
    required: false,
  },
  ];
};

const deleteBooking = function submitDeleteRequestForBookingToServer(booking) {
  // TODO: this. also add a confirmation dialog.
  console.log('deleted.');
  console.log(booking);
};

const onSubmit = function editBookingOnSubmit() {
  console.log('Submitted!');
};

const EditBookingOverlay = ({ booking, exitEditOverlay }) => {
  const [displayDelete, setDisplayDelete] = useState(false);

  if (displayDelete) {
    return (
      <DeleteConfirmation
        booking={booking}
        deleteBooking={() => {
          deleteBooking(booking);
          exitEditOverlay();
        }}
        cancelDelete={() => setDisplayDelete(false)}
      />
    );
  }
  return (
    <div className={style.container}>
      <div className={style.formContainer}>
        <Form
          inputs={getInputs(booking)}
          onSubmit={(event) => onSubmit(event)}
          submitLabel="Save Booking"
        />
        <button type="button" onClick={() => setDisplayDelete(true)}>Delete this booking</button>
        <button type="button" onClick={() => exitEditOverlay()}>Cancel</button>

      </div>
    </div>
  );
};

EditBookingOverlay.propTypes = {
  booking: PropTypes.shape({
  }).isRequired,
  exitEditOverlay: PropTypes.func.isRequired,
};

export default EditBookingOverlay;
