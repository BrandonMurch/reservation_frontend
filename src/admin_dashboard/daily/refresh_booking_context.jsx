import React, { createContext, useContext, useRef } from 'react';
import PropTypes from 'prop-types';

const RefreshDailyBookingContext = createContext();

const RefreshDailyBookingContextProvider = function BoilerPlateForTokenContextProvider({
  refreshFunction,
  children,
}) {
  const func = useRef();
  func.current = refreshFunction;

  const tokenObject = {
    refresh: () => {
      func.current();
    },
  };

  return (
    <RefreshDailyBookingContext.Provider value={tokenObject}>
      {children}
    </RefreshDailyBookingContext.Provider>
  );
};

RefreshDailyBookingContextProvider.propTypes = {
  refreshFunction: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
};

const useRefreshContext = function CustomHookToConsumeTokenContext() {
  return useContext(RefreshDailyBookingContext);
};

export { RefreshDailyBookingContext, useRefreshContext, RefreshDailyBookingContextProvider };
