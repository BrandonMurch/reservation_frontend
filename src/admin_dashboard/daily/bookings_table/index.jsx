// Dependencies
import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import bookingQuickSort from '../bookingQuickSort';
import types from './edit_booking/window_types';

// Components
import EditBookingOverlay from './edit_booking';
import TableRow from './table_row';

// Stylesheets
import style from './bookings_table.module.css';

const loadBookingsIntoMap = function loadBookingsIntoMapByHour(bookings) {
  const bookingsByHourMap = new Map();
  bookings.forEach((booking) => {
    /* REGEX: get index 11 + 12, if 11 is 0, get only index 12
    example 2020-10-10T18:00:00 would get 18, but if it was T08:00:00, it would only get 8;
    */
    const getHourInIsoPattern = '([0-9-T]{11})([1-9])([0-9])';
    const bookingHourMatch = booking.startTime.match(getHourInIsoPattern);
    const bookingHour = bookingHourMatch[2] + bookingHourMatch[3];
    let bookingList;
    if (bookingsByHourMap.has(bookingHour)) {
      bookingList = bookingsByHourMap.get(bookingHour);
    } else {
      bookingList = [];
    }
    bookingList.push(booking);
    bookingsByHourMap.set(bookingHour, bookingList);
  });
  return bookingsByHourMap;
};

const HourRow = ({ hour }) => (
  <tr className={style.hourRow}>
    <td><h2>{`${hour}:00`}</h2></td>
  </tr>
);

HourRow.propTypes = {
  hour: PropTypes.number.isRequired,
};

const tableBookings = function placeBookingsIntoHourSlotsInTable(
  bookingsMap, bookingToEdit, setBookingOverlayWindow,
) {
  const bookingDisplay = [];

  for (let hour = 0; hour < 24; hour++) {
    const hourString = hour.toString();
    if (bookingsMap.has(hourString)) {
      const bookings = bookingsMap.get(hourString);
      bookingQuickSort(bookings);
      bookingDisplay.push(<HourRow key={hour} hour={hour} />);
      bookings.forEach((booking) => {
        const key = booking.startTime + booking.user.username;
        bookingDisplay.push(
          <TableRow
            key={key}
            booking={booking}
            setBookingForEditOverlay={(selectedBooking) => {
              bookingToEdit.current = selectedBooking;
              setBookingOverlayWindow(types.EDIT);
            }}
          />,
        );
      });
    }
  }

  return bookingDisplay;
};

const Bookings = function BookingsTableByHour({ bookings, toggleBookingRefresh }) {
  const [errorBanner, setErrorBanner] = useState(null);
  const [bookingOverlayWindow, setBookingOverlayWindow] = useState(null);
  const selectedBooking = useRef();
  let bookingsTableComponents;
  if (bookings != null && bookings.length !== 0) {
    const bookingsMap = loadBookingsIntoMap(bookings);
    bookingsTableComponents = tableBookings(bookingsMap, selectedBooking, setBookingOverlayWindow);
  }
  return (
    <>
      {errorBanner}
      {bookingOverlayWindow
      && (
      <EditBookingOverlay
        booking={selectedBooking.current}
        entryWindow={bookingOverlayWindow}
        exit={() => {
          selectedBooking.current = null;
          setBookingOverlayWindow(null);
          toggleBookingRefresh();
        }}
        setErrorBanner={setErrorBanner}
      />
      )}

      <div className={style.tableContainer}>
        <table className={style.table}>
          <thead>
            <tr>
              <th>Time</th>
              <th>Table</th>
              <th>Party Size</th>
              <th>Contact</th>
              <th>Comments</th>
              <th><button type="button" onClick={() => setBookingOverlayWindow(types.CREATE)}>Make a booking</button></th>
            </tr>
          </thead>
          <tbody>
            {bookingsTableComponents}
          </tbody>
        </table>
      </div>
    </>
  );
};

Bookings.propTypes = {
  bookings: PropTypes.arrayOf(Object).isRequired,
  toggleBookingRefresh: PropTypes.func.isRequired,
};

export default Bookings;
