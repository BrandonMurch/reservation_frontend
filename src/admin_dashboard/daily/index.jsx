// Dependences
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import useFetch from 'shared/useFetch';
import useTimeHandler from 'shared/useTimeHandler';
import moment from 'moment';
import { Redirect, useParams } from 'react-router-dom';
import { useTokenContext } from 'contexts/token_context';
import { RefreshDailyBookingContextProvider } from './refresh_booking_context';

// Components
import Header from 'general_components/calendar/header';
import Bookings from './bookings_table';

// Style sheets
import style from './daily.module.css';

const DailyController = function ControlDailyDate() {
  const { date } = useParams();
  const timeHandler = useTimeHandler(date);
  const dateString = timeHandler.dateObject.format('yyyy-MM-DD');
  if (date !== dateString) {
    return <Redirect to={`/admin/daily/${dateString}`} />;
  }

  return <Daily {...timeHandler} />;
};

const Daily = function DisplayDailyReservations({ dateObject, dispatchDate }) {
  const token = useTokenContext.getToken;
  const dateString = dateObject.format('yyyy-MM-DD');
  const path = `/bookings?date=${dateString}`;
  const [fetchToggle, toggleFetch] = useState(false);
  const { alternativeRender, response, status } = useFetch(path, { headers: { authorization: `Bearer: ${token}` } }, fetchToggle);
  if (status === 401) {
    return <Redirect to="/admin-login" />;
  }
  if (alternativeRender) {
    return alternativeRender;
  }
  return (
    <RefreshDailyBookingContextProvider refreshFunction={() => toggleFetch(!fetchToggle)}>
      <div className={style.container}>
        <Header
          date={dateObject.format('dddd MMMM Do[,] YYYY')}
          prev={() => {
            dispatchDate({ type: 'prev', unit: 'day' });
          }}
          next={() => {
            dispatchDate({ type: 'next', unit: 'day' });
          }}
          isThisToday={dateObject.startOf('day').isSame(moment().startOf('day'))}
          goToToday={() => dispatchDate({ type: 'current' })}
          dateObject={dateObject}
          dispatchDate={dispatchDate}
        />
        <Bookings bookings={response} date={dateString} />
      </div>
    </RefreshDailyBookingContextProvider>
  );
};

Daily.propTypes = {
  dispatchDate: PropTypes.func.isRequired,
  dateObject: PropTypes.shape({
    format: PropTypes.func.isRequired,
    startOf: PropTypes.func.isRequired,
  }).isRequired,
};

export default DailyController;
