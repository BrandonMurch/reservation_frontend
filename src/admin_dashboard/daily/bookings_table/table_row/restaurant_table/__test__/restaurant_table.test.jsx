// Dependencies
import React from 'react';
import {
  render, fireEvent, screen, waitForElementToBeRemoved,
} from '@testing-library/react';
import { create, act } from 'react-test-renderer';
import { exampleBooking1, tables } from 'test_utils/examples';
import { mockFetch } from 'test_utils';
import { RefreshDailyBookingContextProvider } from 'admin_dashboard/daily/refresh_booking_context';

// Components
import RestaurantTable from '../index';
import userEvent from '@testing-library/user-event';

describe('<RestaurantTable />', () => {
  let fetchSpy;
  const mockRefresh = jest.fn();
  const props = {
    booking: exampleBooking1,
  };
  beforeEach(async () => {
    fetchSpy = jest.spyOn(global, 'fetch').mockImplementation(() => mockFetch(200, tables));
    render(
      <RefreshDailyBookingContextProvider refreshFunction={() => mockRefresh}>
        <RestaurantTable {...props} />
        ,
      </RefreshDailyBookingContextProvider>,
    );

    await waitForElementToBeRemoved(() => screen.queryByRole('presentation'));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', async () => {
    let tree;
    await act(async () => {
      tree = await create(
        <RefreshDailyBookingContextProvider refreshFunction={() => mockRefresh}>
          <RestaurantTable {...props} setIsLoading={jest.fn()} tableList={tables} />
          ,
        </RefreshDailyBookingContextProvider>,
      );
    });

    expect(tree).toMatchSnapshot();
  });
  it('should display the table within the input box', () => {
    const input = screen.getByRole('textbox', { name: 'Restaurant Table' });
    expect(input.value).toEqual('2');
  });
  it('should call fetch on blur', async () => {
    const input = screen.getByRole('textbox', { name: 'Restaurant Table' });
    userEvent.type(input, 'a');
    fireEvent.blur(input);

    await waitForElementToBeRemoved(() => screen.queryByLabelText('loading'));
    expect(fetchSpy).toHaveBeenCalledTimes(2);
  });
});
