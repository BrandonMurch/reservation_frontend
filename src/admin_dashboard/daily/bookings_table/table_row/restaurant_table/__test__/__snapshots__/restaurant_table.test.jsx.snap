// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RestaurantTable /> should match snapshot 1`] = `
Array [
  <div
    className="container"
  >
    <div
      className="inputGroup"
    >
      <label
        className="hiddenLabelText"
        htmlFor="table"
      >
        Restaurant Table
      </label>
      <input
        className="input"
        id="table"
        name="table"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        pattern={null}
        placeholder="Restaurant Table"
        required={false}
        type="text"
        value="2"
      />
    </div>
    
  </div>,
  ",",
]
`;
